### PR TITLE
[WIP] Use visitor based file system mirror

### PR DESCRIPTION
**Pull request by wolfs**
Instead of returning a full list of files in a directory tree, we return something which can be visited.
This allows for changing the file system mirror to a different data structure where all the file collection snapshots are not determined up front.
Only the visitor should then decide how much information it needs.

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/DefaultFileSystemSnapshotter.java [hunk 1]
```diff
@@ -154,7 +139,7 @@ public FileTreeSnapshot create() {
      * before any downstream task uses them.
      */
     @Override
-    public FileTreeSnapshot snapshotDirectoryTree(final DirectoryFileTree dirTree) {
+    public VisitableDirectoryTree snapshotDirectoryTree(final DirectoryFileTree dirTree) {
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/DefaultFileSystemSnapshotter.java]: These methods don't snapshot the trees anymore, they just create the "snapshotting visitor".

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/AbstractFileCollectionSnapshotter.java [hunk 2]
```diff
@@ -91,15 +90,14 @@ public void visitCollection(FileCollectionInternal fileCollection) {
 
         @Override
         public void visitTree(FileTreeInternal fileTree) {
-            List<FileSnapshot> descendants = fileSystemSnapshotter.snapshotTree(fileTree);
-            fileSnapshotVisitor.visitFileTreeSnapshot(descendants);
+            VisitableDirectoryTree treeSnapshot = fileSystemSnapshotter.snapshotTree(fileTree);
+            fileSnapshotVisitor.visitFileTreeSnapshot(treeSnapshot);
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/AbstractFileCollectionSnapshotter.java]: Would it make sense to turn this around into `treeSnapshot.visit(...)`?

Also, `treeSnapshot` is not really a snapshot of the tree anymore, it's just the entry point you can use to traverse from. I think we should rename it accordingly.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/FileSystemSnapshotter.java [hunk 3]
```diff
@@ -55,20 +55,14 @@
...
-     * Returns the current snapshot of the contents and meta-data of the given directory.
-     * The provided directory must exist and be a directory.
-     */
-    FileTreeSnapshot snapshotDirectoryTree(File dir);
-
     /**
      * Returns the current snapshot of the contents and meta-data of the given directory tree.
      */
-    FileTreeSnapshot snapshotDirectoryTree(DirectoryFileTree dirTree);
+    VisitableDirectoryTree snapshotDirectoryTree(DirectoryFileTree dirTree);
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/FileSystemSnapshotter.java]: We also need to update the Javadocs here.

**Comment by wolfs** [comment]
I ran a performance test on the branch: https://builds.gradle.org/viewLog.html?buildId=13332106&buildTypeId=Gradle_Check_PerformanceTestCoordinator&tab=report_project944_Performance

Nothing too surprising so far. The large java Multiproject build has the biggest regressions.

**Comment by wolfs** [comment]
Closing as superseded by #5661.

**Pull request closed without merging**